File: app/src/main/java/be/digitalia/fosdem/fragments/PersonsListFragment.java
package be.digitalia.fosdem.fragments;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.util.ObjectsCompat;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.paging.PagedList;
import androidx.paging.PagedListAdapter;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import be.digitalia.fosdem.R;
import be.digitalia.fosdem.activities.PersonInfoActivity;
import be.digitalia.fosdem.adapters.SimpleItemCallback;
import be.digitalia.fosdem.model.Person;
import be.digitalia.fosdem.viewmodels.PersonsViewModel;

public class PersonsListFragment extends RecyclerViewFragment implements Observer<PagedList<Person>> {

	private PersonsAdapter adapter;

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		adapter = new PersonsAdapter();
	}

	@NonNull
	@Override
	protected RecyclerView onCreateRecyclerView(LayoutInflater inflater, ViewGroup container, @Nullable Bundle savedInstanceState) {
		return (RecyclerView) inflater.inflate(R.layout.recyclerview_fastscroll, container, false);
	}

	@Override
	protected void onRecyclerViewCreated(RecyclerView recyclerView, Bundle savedInstanceState) {
		recyclerView.setLayoutManager(new LinearLayoutManager(recyclerView.getContext()));
		recyclerView.addItemDecoration(new DividerItemDecoration(recyclerView.getContext(), DividerItemDecoration.VERTICAL));
	}

	@Override
	public void onActivityCreated(Bundle savedInstanceState) {
		super.onActivityCreated(savedInstanceState);

		setAdapter(adapter);
		setEmptyText(getString(R.string.no_data));
		setProgressBarVisible(true);

		final PersonsViewModel viewModel = new ViewModelProvider(this).get(PersonsViewModel.class);
		viewModel.getPersons().observe(getViewLifecycleOwner(), this);
	}

	@Override
	public void onChanged(PagedList<Person> persons) {
		adapter.submitList(persons);
		setProgressBarVisible(false);
	}

	private static class PersonsAdapter extends PagedListAdapter<Person, PersonViewHolder> {

		private static final DiffUtil.ItemCallback<Person> DIFF_CALLBACK = new SimpleItemCallback<Person>() {
			@Override
			public boolean areContentsTheSame(@NonNull Person oldItem, @NonNull Person newItem) {
				return ObjectsCompat.equals(oldItem.getName(), newItem.getName());
			}
		};

		PersonsAdapter() {
			super(DIFF_CALLBACK);
		}

		@NonNull
		@Override
		public PersonViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
			View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.simple_list_item_1_material, parent, false);
			return new PersonViewHolder(view);
		}

		@Override
		public void onBindViewHolder(@NonNull PersonViewHolder holder, int position) {
			final Person person = getItem(position);
			if (person == null) {
				holder.clear();
			} else {
				holder.bind(person);
			}
		}
	}

	static class PersonViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
		final TextView textView;

		Person person;

		PersonViewHolder(@NonNull View itemView) {
			super(itemView);
			textView = itemView.findViewById(android.R.id.text1);
			itemView.setOnClickListener(this);
		}

		void clear() {
			this.person = null;
			textView.setText(null);
		}

		void bind(@NonNull Person person) {
			this.person = person;
			textView.setText(person.getName());
		}

		@Override
		public void onClick(View view) {
			if (person != null) {
				final Context context = view.getContext();
				Intent intent = new Intent(context, PersonInfoActivity.class)
						.putExtra(PersonInfoActivity.EXTRA_PERSON, person);
				context.startActivity(intent);
			}
		}
	}
}

File: app/src/main/kotlin/be/digitalia/fosdem/fragments/PersonsListFragment.kt
package be.digitalia.fosdem.fragments

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.paging.LoadState
import androidx.paging.PagingDataAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import be.digitalia.fosdem.R
import be.digitalia.fosdem.activities.PersonInfoActivity
import be.digitalia.fosdem.adapters.createSimpleItemCallback
import be.digitalia.fosdem.model.Person
import be.digitalia.fosdem.utils.launchAndRepeatOnLifecycle
import be.digitalia.fosdem.viewmodels.PersonsViewModel
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

@AndroidEntryPoint
class PersonsListFragment : Fragment(R.layout.recyclerview_fastscroll) {

    private val viewModel: PersonsViewModel by viewModels()

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val adapter = PersonsAdapter()
        val holder = RecyclerViewViewHolder(view).apply {
            recyclerView.apply {
                layoutManager = LinearLayoutManager(context)
                addItemDecoration(DividerItemDecoration(context, DividerItemDecoration.VERTICAL))
            }
            setAdapter(adapter)
            emptyText = getString(R.string.no_data)
            isProgressBarVisible = true
        }

        viewLifecycleOwner.lifecycleScope.launch {
            adapter.loadStateFlow.first { it.refresh !is LoadState.Loading }
            holder.isProgressBarVisible = false
        }

        viewLifecycleOwner.launchAndRepeatOnLifecycle {
            viewModel.persons.collectLatest { pagingData ->
                adapter.submitData(pagingData)
            }
        }
    }

    private class PersonsAdapter : PagingDataAdapter<Person, PersonViewHolder>(DIFF_CALLBACK) {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PersonViewHolder {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.simple_list_item_1_material, parent, false)
            return PersonViewHolder(view)
        }

        override fun onBindViewHolder(holder: PersonViewHolder, position: Int) {
            val person = getItem(position)
            if (person == null) {
                holder.clear()
            } else {
                holder.bind(person)
            }
        }

        companion object {
            private val DIFF_CALLBACK: DiffUtil.ItemCallback<Person> = createSimpleItemCallback { it.id }
        }
    }

    private class PersonViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        val textView: TextView = itemView.findViewById(android.R.id.text1)

        var person: Person? = null

        init {
            itemView.setOnClickListener(this)
        }

        fun clear() {
            person = null
            textView.text = null
        }

        fun bind(person: Person) {
            this.person = person
            textView.text = person.name
        }

        override fun onClick(view: View) {
            person?.let {
                val context = view.context
                val intent = Intent(context, PersonInfoActivity::class.java)
                        .putExtra(PersonInfoActivity.EXTRA_PERSON, it)
                context.startActivity(intent)
            }
        }
    }
}